plugins/modules/rpm_repository.py

```diff
@@ -138,9 +138,9 @@ def main():
             desired_attributes["remote"] = remote.href
 
         repo_config = module.params["repo_config"]
-        # Encode the repo_config unless its a string, then assume it is pre-formatted JSON
-        if repo_config is not None and not isinstance(repo_config, string_types):
-            desired_attributes["repo_config"] = json.dumps(repo_config)
+        # Encode the repo_config if its a pre-formatted JSON string
+        if repo_config is not None and isinstance(repo_config, string_types):
+            desired_attributes["repo_config"] = json.loads(repo_config)
```

Comment:
What happens if it's not a string?
I'm unsure how this is supposed to be represented. It looks like we are missing either case in both versions. We should have a test for this.

"A JSON document or data structure describing a config.repo file"

Reply:
1. As a string, it's converted to a json blob
      `repo_config: '{"gpgcheck": 1}'`

2. As a hash of key:pairs, ansible properly converts this to a json blob and passes to the API.
```
      repo_config:
        gpgcheck: 1
        anotherkey: value
        anotherkey2: value
```

In either of these cases, it'll work. The current code test for the (1) string value, should we have a test for (2) as well?

Reply:
ATM desired_attributes is only set on one branch of the if.
So yes, I could really be convinced by seeing both versions in a test.

Reply:
@mdellweg - These tests are already in place, just the fixtures were not proper. (I'm now remembering)
See: https://github.com/pulp/squeezer/blob/develop/tests/playbooks/rpm_repository.yaml#L115-L161

>> desired_attributes is only set on one branch of the if

desired_attributes is initially set with this code:
https://github.com/pulp/squeezer/blob/develop/plugins/modules/rpm_repository.py#L127-L129

The IF statement is modifying the `desired_attributes["repo_config"]` to correctly format the JSON blob when a string vs dict is provided.

Reply:
I see.
Then let's make this more obvious and fix `desired_attributes["repo_config"]` in place without reading from `module.params` again.

Reply:
It would also be ok to just require it to be parsed from json already. I think there is a "from_json" filter in ansible to bridge the gap.

Reply:
Removed the rewriting from module.params and in place fix for desired_attributes.

Reply:
Can you change the test in a way, that the module is faced with both versions once?

Reply:
I've changed the check to test the json version and then the string version, validating no changes to the repo are detected. This confirms different endpoints produce the same output.